Label list filter

agrogers opened the pull request:
I have this mostly working now. But the code feels really messy, even by my pretty low standards. I will come back to it later and see if i can improve it. If anyone has some helpful suggestions on either interface or code, please let me know. This PR should close https://github.com/AndBible/and-bible/issues/1620

Here is how it works:
- It starts with an 'ALL' button and a magnifing search button
- Pushing the search button toggles the visibiluty of the search textbox and buttons
- The left side button allows you to choose to match either on the start of the label or on any part inside the label (shown as a blue button)
- The Save button will save the text as a button in the row at the top
- It will remember the type of search and display the 'match any part' of the label button with a blue border
- Long-pressing a search button will delete the button
-  The filter seems to work nicely on all the dialogs that use a label list

![image](https://user-images.githubusercontent.com/13920678/150124666-2e079765-b80e-4933-8826-dd0ac5581890.png)

![image](https://user-images.githubusercontent.com/13920678/150124747-53f65f7d-c7a5-4581-a017-c3f52738e1b5.png)

agrogers committed:
Added fixed buttons to filter label list

agrogers committed:
Pretty much working now.

agrogers committed:
Fixed some text and button refresh/redraw issues

agrogers committed:
Added better regex support and bug fixes

agrogers committed:
Update strings.xml

agrogers committed:
Added toJson functions.

agrogers committed:
Implemented json loading of saved searches

agrogers commented:
Button text shows fine now in dark mode. Options are serialized and saved correctly. I believe this is ready for release now. I don't plan to do anymore on it for a bit. And might I say I couldn't live without it now :)

![image](https://user-images.githubusercontent.com/13920678/156874015-d92d8625-c900-4d15-a3b5-df55197e29ba.png)

agrogers committed:
Merge branch 'develop' into Label-List-Filter

tuomas2 commented:
Then push that "Ready for review" button :-)

agrogers commented:
> Then push that "Ready for review" button :-)

I should merge the latest 'develop' first. I will do that soon.

agrogers commented:
OK, develop has been merged.

agrogers committed:
Added check for existing installs.

tuomas2 committed:
Merge branch 'develop' into agrogers/Label-List-Filter

tuomas2 commented:
- [ ] save button looks like toggle button (confusing)
- [ ] 3rd mode: "regex" to ab-button? and explain in help
- [x] Other modes to ignore regex.
- [x] filtering finds always "Unlabeled"
- [x] Help screen (?-button) should instruct to use filter (for example long-pressing removes saved items, which can't be easily guessed)
- [ ] layout does not work in RTL 
- [x] wont fix ~perhaps all-button is not needed. To disable filter, just tap button again. Or replace with x-button on the right? search button (magnifier glass) should probably be the leftmost button.~
- [x] ~search row should not be visible by default ever~ I'm okay with leaving it as saved setting
- [x] ~check: using fromDb = true in updateLabelList will reload labels from db, i.e. changes will be lost?~ apparently not so, but code needs to be clarified still (fromDb is misleading)
- [x] perhaps show `*`-markings in search buttons (like `cat* ` or `*cat*` etc). I don't think blue border color is very clear indicator.

tuomas2 commented on app/src/main/java/net/bible/android/view/activity/bookmark/ManageLabelItemAdapter.kt:
```diff
@@ -72,10 +72,6 @@ class ManageLabelItemAdapter(context: Context?,
                 if (data.showCheckboxes) {
                     checkbox.setOnCheckedChangeListener { _, isChecked ->
                         if (isChecked) {
-                            if (!manageLabels.selectMultiple) {
-                                data.selectedLabels.clear()
-                                data.bookmarkPrimaryLabel = null
-                            }
```
why this change?

agrogers replied:
I think this was a remnant of when you could choose to select either single or multiple labels. Didn't you hide that option and then no one complained about it so now we can only choose multiple labels. Thus that code is no longer needed.

tuomas2 commented on app/src/main/res/values/strings.xml:
```diff
@@ -406,6 +406,8 @@
     <string name="auto_assign_labels_primary">Auto-assign as primary</string>
     <string name="label_color">Label color</string>  <!-- can be removed -->
     <string name="label_color_example">Example</string>
+    <string name="match_any_text">*ab*</string>  <!-- Probably does not need to be translated. They are symbols for a button image -->
+    <string name="match_start_of_text">Ab*</string>  <!-- Probably does not need to be translated. They are symbols for a button image -->
```
-> untranslated_strings.xml

tuomas2 replied:
though, we have also rtl languages, that should be taken care of. Perhaps by keeping then translatable....

agrogers replied:
I don't think of these as strings but as an icon. I am not sure that there would be a translatable phrase for them. The important character is the *

tuomas2 replied:
The issue is that on RTL languages, * should be on the other side than what we think.

agrogers commented:
> * [ ]  save button looks like toggle button (confusing)

Feel free to change it.

> * [ ]  filtering finds always "Unlabeled"

Not on my phone. Can't recall what it did with a default install.

tuomas2 commented:
> Not on my phone. Can't recall what it did with a default install.

It is shown in some views, and not others. See for example "label settings" view. 

PS. I'm trying to work on this PR today, let's see how far I get.

tuomas2 committed:
Cleanup & improve code

Mostly IDE suggestions

tuomas2 committed:
Layout improvements

tuomas2 commented on app/src/main/java/net/bible/android/view/activity/bookmark/ManageLabels.kt:
```diff
@@ -79,15 +99,203 @@ class ManageLabels : ListActivityBase() {
+                newButton.tag = searchOption.id
+                newButton.setBackgroundResource(R.drawable.button_filter)
+                newButton.layoutParams = ViewGroup.LayoutParams(LinearLayout.LayoutParams.WRAP_CONTENT, 60)
+                newButton.minWidth = 80
```
- [x] this and other definitions are pixels if set from code. Should be dp or dip. I think they are not equal.

tuomas2 committed:
Improve code

tuomas2 commented on app/src/main/java/net/bible/android/view/activity/bookmark/ManageLabels.kt:
```diff
@@ -538,21 +783,40 @@ class ManageLabels : ListActivityBase() {
+                filterText = "^[" + filterText.takeLast(filterText.length-1) + "]"
+            }
+            // TODO: too wide catch!
+            val regex = try {filterText.toRegex(RegexOption.IGNORE_CASE)} catch (e:Exception) {"".toRegex()}
```
- [x] catch too wide

tuomas2 commented on app/src/main/java/net/bible/android/view/activity/bookmark/ManageLabels.kt:
```diff
@@ -581,11 +845,15 @@ class ManageLabels : ListActivityBase() {
+            // Some sanity check. These checks clear the changed state of the labels.
+            // AGR: I don't know whether these are really needed. But having them clear these settings when the filter is applied causes me trouble.
+//            data.autoAssignLabels.myRemoveIf { !labelIds.contains(it) }
+//            data.favouriteLabels.myRemoveIf { !labelIds.contains(it) }
```
- [x] TODO

tuomas2 committed:
Improve code

tuomas2 committed:
Fix isAllcaps

agrogers committed:
Update ManageLabels.kt

This is needed because the keyboard disappears after each key stroke. Problem happened on emulator and device. On emulator it is obviously but not a problem because the computer keyboard keeps working.

agrogers commented:
I have added the HideKeyboard flag back in because I believe it needs it. Without it the keybaord disappears after each keystroke - both on the emulator and my device.

tuomas2 commented on app/src/main/java/net/bible/android/view/activity/bookmark/ManageLabels.kt:
```diff
@@ -70,6 +87,23 @@ fun WorkspaceEntities.WorkspaceSettings.updateFrom(resultData: ManageLabels.Mana
+) {
+    // TODO: Presently 'name' is not used. I want to show a dialog on the long press of a button
+    //  that allows the deleting or renaming of a button. This allows for complex regex strings but with a more human readable name
+    init {
```
Removing name for now. It can be added in the future if functionality is added.

tuomas2 committed:
Improve code

tuomas2 committed:
Code improvements

tuomas2 committed:
Fix build

agrogers committed:
Update ManageLabelItemAdapter.kt

Reclaim the wasted space which usually  shows an empty column for the Primary icon.

agrogers commented:
Last pullrequest removes the wasted space due to the almost always empty column for the Primary icon.

![image](https://user-images.githubusercontent.com/13920678/163919710-39b1e746-40c3-45d1-b167-fc3f647d6be9.png)

tuomas2 commented on app/src/main/java/net/bible/android/view/activity/bookmark/ManageLabels.kt:
```diff
@@ -79,15 +115,184 @@ class ManageLabels : ListActivityBase() {
+
+    private fun loadFilteringSettings() {
+        searchInsideTextButtonActive = CommonUtils.settings.getBoolean("labels_list_filter_searchInsideTextButtonActive", false)
+        showTextSearch = CommonUtils.settings.getBoolean("labels_list_filter_showTextSearch", false)
```
Do we want that this setting is remembered? I would assume people do not want to have this on, unless they tap magnifier glass icon.

agrogers replied:
I almost always have the text search enabled. I find that just as fast or faster than using the buttons :) Should have just done that in the first place!

agrogers replied:
I almost always have the text search enabled. I find that just as fast or faster than using the buttons :) Should have just done that in the first place!

tuomas2 commented on app/src/main/java/net/bible/android/view/activity/bookmark/ManageLabels.kt:
```diff
@@ -79,15 +115,184 @@ class ManageLabels : ListActivityBase() {
+        showTextSearch = CommonUtils.settings.getBoolean("labels_list_filter_showTextSearch", false)
+        val options = CommonUtils.settings.getString("labels_list_filter_searchTextOptions") ?: ""
+        searchOptionList.clear()
+        if (options.isNotEmpty() && options.first() == '[') {
```
why the second condition?

tuomas2 replied:
improved

tuomas2 committed:
Code improvements

tuomas2 committed:
Merge branch 'develop' into agrogers/Label-List-Filter

tuomas2 committed:
Improve code

tuomas2 committed:
Fix save button

tuomas2 committed:
Fix save button

tuomas2 committed:
Improvements. Different visualization of asterisk

tuomas2 committed:
Fix adding / removing btns

tuomas2 committed:
Merge branch 'develop' into agrogers/Label-List-Filter

tuomas2 commented on app/src/main/java/net/bible/android/view/activity/bookmark/ManageLabels.kt:
```diff
@@ -538,21 +747,40 @@ class ManageLabels : ListActivityBase() {
+            // Exclude the 'unlabeled label' but include all selected labels plus those that match the filter clause.
+            // We always want to see the labels we have selected even when filtering.
+            shownLabels.addAll(allLabels.filter {
+                !it.isUnlabeledLabel and (
```
`and` and `or` are bitwise operators in kotlin. Use && and ||.

tuomas2 replied:
fixed

tuomas2 committed:
Code improvements. Simplify features a bit too to make simpler code...

tuomas2 committed:
Escape all regex inside searchText

tuomas2 committed:
Clean up code a bit

tuomas2 commented:
> [ ] 3rd mode: "regex" to ab-button? and explain in help

What do you think, would full 'regex mode' be needed? Would be used by geeks only, of course... Now all regex will be escaped, so patterns can't be used (except those simple ones provided by `*Ab*/Ab*` button).

tuomas2 commented:
> Last pullrequest removes the wasted space due to the almost always empty column for the Primary icon.
> 
> ![image](https://user-images.githubusercontent.com/13920678/163919710-39b1e746-40c3-45d1-b167-fc3f647d6be9.png)

~Side effect is that now it is more difficult to change primary label. (can't be done from labels listing)~ no, it is possible (because all selected will have it visible)

tuomas2 commented:
RTL now:

![image](https://user-images.githubusercontent.com/5811789/166096626-d52a46ef-2b6a-429f-a073-8fa71b0dba85.png)

tuomas2 committed:
Improve code

agrogers commented:
> > [ ] 3rd mode: "regex" to ab-button? and explain in help
> 
> What do you think, would full 'regex mode' be needed? Would be used by geeks only, of course... Now all regex will be escaped, so patterns can't be used (except those simple ones provided by `*Ab*/Ab*` button).
> 
> 

If I could name an entry I would use regex. But I agree only feels would use it. Once it was practical I thought I would add some regex strings I thought popular to the manual or popup help and then people just had to copy and paste.

agrogers commented:
Does the modifications show the * for all modes or just the new mode? I used the color coding of buttons because I was trying to keep the buttons as small as possible. I could almost get A through Z on two lines.

tuomas2 commented:
> Does the modifications show the * for all modes or just the new mode? I used the color coding of buttons because I was trying to keep the buttons as small as possible. I could almost get A through Z on two lines.

I think color coding does not communicate user what is going on. I prefer wasting that space because user understands it better.

agrogers commented:
That's OK. I can change it back on my build if I don't like it 🙂

But I think it is wasted space - repeating the same thing 26 times in
limited space is usually a indicator to me that there is a better way of
doing it. I would put this in the category of 'let the interface teach the
user' rather than spelling it out for them.

On Sun, 1 May 2022, 1:56 am Tuomas Airaksinen, ***@***.***>
wrote:

> Does the modifications show the * for all modes or just the new mode? I
> used the color coding of buttons because I was trying to keep the buttons
> as small as possible. I could almost get A through Z on two lines.
>
> I think color coding does not communicate user what is going on. I prefer
> wasting that space because user understands it better.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/AndBible/and-bible/pull/2058#issuecomment-1114041180>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADKGTJUWEIVFJNG6BZYWLCDVHWCPBANCNFSM5MJXE3PQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

tuomas2 committed:
Merge branch 'develop' into agrogers/Label-List-Filter

tuomas2 committed:
Remove (probably) unnecessary sanity checks

tuomas2 committed:
Improve & simplify code

tuomas2 committed:
Improve & simplify code

tuomas2 committed:
Add help text

tuomas2 committed:
Reset filter by pressing button again

tuomas2 committed:
Remove "all" button

tuomas2 committed:
Improve help text

tuomas2 committed:
Fix hide empty "active labels" category

tuomas2 committed:
Remove redudant comment

tuomas2 commented:
I think this is now good (enough) to go. Let's see if beta testing will give reasons for improving this further. 

Very convenient and nice feature! Thank you.
